### ISSUE TITLE
much lower mAP than reported with 4 GPUs

### ISSUE THREAD

**Issue by youthHan** [jd730/OICR-pytorch#23]
Thank you for the work of the reimplementation.

I have tried to run the code on pascal voc 2007, however got a much lower mAP 36.27 that reported.

AP list:
AP for aeroplane = 0.5698
AP for bicycle = 0.5532
AP for bird = 0.2643
AP for boat = 0.1777
AP for bottle = 0.1356
AP for bus = 0.5892
AP for car = 0.5607
AP for cat = 0.3745
AP for chair = 0.2180
AP for cow = 0.3692
AP for diningtable = 0.3320
AP for dog = 0.2192
AP for horse = 0.2363
AP for motorbike = 0.5885
AP for person = 0.1010
AP for pottedplant = 0.2027
AP for sheep = 0.3001
AP for sofa = 0.4573
AP for train = 0.4927
AP for tvmonitor = 0.5130
Mean AP = 0.3627

The script I use is:
`CUDA_VISIBLE_DEVICES=0,1,2,3 python3 trainval_net.py --dataset pascal_voc \
--net vgg16 --bs 8 --nw 12 --save_dir 'results/voc_multi_train_ilabel_full' \
--lr 0.001 --epochs 7 --lr_decay_step 4 \
--cuda --disp_interval 50 \
--vis --mGPUs --checkpoint_interval 1750 \
--model 'oicr' --threshold 0.1 `

I changed the lr-decay iteration from `4e4` to dynamically  `lr_decay_step * iterations_per_epoch` in `trainval_net.py`. 

Thank you for the help in advance.

**Comment by youthHan** [comment]
Besides, I have noticed the former issues #20 #17 and my setting is using ROIPool as default.

**Comment by jd730** [comment]
Hi @youthHan, 
Thank you for your interest in the repository. Unfortunately, I only considered the single GPU environment when I made this repo so it is possible to have any unexpected problem in the code.
Moreover, lr_decay in different steps can affect the final accuracy.

Could you first test 
```
CUDA_VISIBLE_DEVICES=0,1 python3 trainval_net.py --dataset pascal_voc \ --net vgg16 --bs 4 --nw 12 --save_dir 'results/voc_multi_train_ilabel_full' \ --lr 0.002 --lr_decay_step 4 \ --cuda --disp_interval 50 \ --vis --mGPUs --checkpoint_interval 1750 \ --model 'oicr' --threshold 0.1
```
with lr decay at 40k iterations and deleting https://github.com/jd730/OICR-pytorch/blob/bcfe9cd2d7ab77fc46bc031da8fd27bd9971d57d/trainval_net.py#L408

It might be a good starting point to check whether there is no error or not.
(Since current code does forward twice and backward, it should be same as 2 GPUs with twice learning rate)

**Comment by youthHan** [comment]
Thanks for your suggestion @jd730 . 

And I have experimented with default configurations and adjust the learning rate to 0.004 on 4 GPUs. The final mAP is 33.45 compared with 36.27 mentioned before.

The results following your suggestion will be updated once available.

**Comment by youthHan** [comment]
Hi, @jd730 
The mAP is 34.26. I'm confused about why it is.

**Comment by jd730** [comment]
Hi @youthHan ,
Did you get the correct score when you run on a single GPU?
Moreover, did you change the code to use `lr_decay_steps`? I just checked that in my implementation, although there is a flag, it is not used since I manually set when to decay on  https://github.com/jd730/OICR-pytorch/blob/bcfe9cd2d7ab77fc46bc031da8fd27bd9971d57d/trainval_net.py#L404

**Comment by youthHan** [comment]
I can get the correct mAP 41.7 on a single GPU.
In the experiment without accumulated backward on 2 GPUs and the one with single GPU, lr decays at 4e4 iterations and ends at 7e4. 

And I have changed the code to use `lr_decay_steps` in the experiments below.
In the experiment with the normal setting on 4GPUs as mentioned in the first comment, lr decays at epoch 4 and ends at epoch 7.  After training iterations increased to a total 10 epochs and `lr_decay_steps` as 8 epochs, I got improved mAP as 39.7 compared to 36.27 before.

I'm quite confused about the abnormal. I am not familiar with multi-gpu training code, is it possible something wrong there? Could you please check it or give me a hint?
